File: DFS/RobotMove/1041. Robot Bounded In Circle/1041. Robot Bounded In Circle.cpp
class Solution {
public:
    bool isRobotBounded(string instructions) {
        vector<int>cur = {0, 0};
        vector<vector<int>>dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int dir = 0;// 0 north up, 1 right, 2 down, 3 left;
        
        for ( int i = 0; i < instructions.size(); i++ ) {
            if (instructions[i] == 'G') {
                cur[0] += dirs[dir][0];
                cur[1] += dirs[dir][1];
            } else if (instructions[i] == 'L') {
                dir = (dir+3) %4;
            } else {
                dir = (dir+1) %4;
            }
        }
        
        if (cur[0] == 0 && cur[1] == 0) return true; // locate in the original point
        else if (dir == 0) return false; // north direction but has offset
        return true;
    }
};

File: Greedy/1041. Robot Bounded In Circle/1041. Robot Bounded In Circle.java
class Solution {
    public boolean isRobotBounded(String instructions) {
        // direction : i%4;
        int x = 0, y = 0, direction = 0;
        for (char c : instructions.toCharArray()) {
            if (c == 'G') {
                if (direction == 0) {
                    y += 1;
                } else if (direction == 1) {
                    x += 1;
                } else if (direction == 2) {
                    y -= 1;
                } else {
                    x -= 1;
                }
            } else if (c == 'L') {
                direction = (direction-1+4)%4;
            } else if (c == 'R') {
                direction = (direction+1+4)%4;
            }
        }
        
        if (x == 0 && y==0) return true;
        else {
            if (direction != 0) return true;
            else return false;
        }
    }
}
